File: 2331-Evaluate-Boolean-Binary-Tree/short-circuit.py
# https://leetcode.com/problems/evaluate-boolean-binary-tree/submissions/1032801460/

class Solution:
    def printLeafs(self, node) -> bool:
        if node == None or node.val <= 1:
            return node.val == 1
        
        left = self.printLeafs(node.left)
        if (node.val == 2 and left) or (node.val == 3 and not left):
            return left
        
        return self.printLeafs(node.right)

    def evaluateTree(self, root: Optional[TreeNode]) -> bool:
        return self.printLeafs(root)

File: 2331-Evaluate-Boolean-Binary-Tree/short-circuit.js
// https://leetcode.com/problems/evaluate-boolean-binary-tree/submissions/1032837328/

function evaluate(node){
    if(node === null || node.val <= 1){
        return node.val === 1
    }

    const left = evaluate(node.left)
    if((node.val === 2 && left) || (node.val === 3 && !left)){
        return left
    }
    return evaluate(node.right)
}

var evaluateTree = function(root) {
    return evaluate(root)
};
